File: 56-merge-intervals/merge-intervals.go
func merge(intervals [][]int) [][]int {
	// Sort intervals by their start value
	sort.Slice(intervals, func(i, j int) bool {
		return intervals[i][0] < intervals[j][0]
	})

	result := [][]int{intervals[0]}

	for i := 1; i < len(intervals); i++ {
		prev := result[len(result)-1]
		curr := intervals[i]

		if curr[0] <= prev[1] {
			// Merge: expand the end if needed
			if curr[1] > prev[1] {
				prev[1] = curr[1]
			}
		} else {
			result = append(result, curr)
		}
	}

	return result
}

File: 56-merge-intervals/merge-intervals.java
class Solution {
    public int[][] merge(int[][] intervals) {

        Arrays.sort(intervals, (a, b) -> Integer.compare(a[0], b[0]));
        List<int[]> list = new ArrayList<>();
        list.add(intervals[0]);
        for (int i = 0; i < intervals.length; i++) {
            int prev[] = list.get(list.size() - 1);
            int curr[] = intervals[i];

            if (curr[0] <= prev[1]) {
                // merge..
                prev[0] = Math.min(prev[0], curr[0]);
                prev[1] = Math.max(prev[1], curr[1]);
            } else {
                list.add(intervals[i]);
            }

        }

        return list.toArray(new int[list.size()][2]);

    }
}